src/vmm/src/vstate/vcpu/mod.rs

```diff
@@ -458,108 +441,115 @@ impl Vcpu {
             self.kvm_vcpu.fd.set_kvm_immediate_exit(0);
             return Ok(VcpuEmulation::Interrupted);
         }
-        match self.emulate() {
-            Ok(run) => match run {
-                VcpuExit::MmioRead(addr, data) => {
-                    if let Some(mmio_bus) = &self.kvm_vcpu.mmio_bus {
-                        let _metric = METRICS.vcpu.exit_mmio_read_agg.record_latency_metrics();
-                        mmio_bus.read(addr, data);
-                        METRICS.vcpu.exit_mmio_read.inc();
-                    }
-                    Ok(VcpuEmulation::Handled)
-                }
-                VcpuExit::MmioWrite(addr, data) => {
-                    if let Some(mmio_bus) = &self.kvm_vcpu.mmio_bus {
-                        let _metric = METRICS.vcpu.exit_mmio_write_agg.record_latency_metrics();
-                        mmio_bus.write(addr, data);
-                        METRICS.vcpu.exit_mmio_write.inc();
-                    }
-                    Ok(VcpuEmulation::Handled)
+
+        match self.kvm_vcpu.fd.run() {
+            Err(ref err) if err.errno() == libc::EINTR => {
+                self.kvm_vcpu.fd.set_kvm_immediate_exit(0);
+                // Notify that this KVM_RUN was interrupted.
+                Ok(VcpuEmulation::Interrupted)
+            }
+            emulation_result => handle_kvm_exit(&mut self.kvm_vcpu.peripherals, emulation_result),
+        }
+    }
+}
+
+/// Handle the return value of a call to [`VcpuFd::run`] and update our emulation accordingly
+fn handle_kvm_exit(
+    emulation_state: &mut Peripherals,
```

Comment:
maybe rename this to `peripherals: &mut Peripherals` for consistency?

Reply:
done :)